### DISCUSSION TITLE
Difference between Default and Implicit DSL

### DISCUSSION THREAD

**Discussion by superflyer11** [thelfer/tfel#630] [Q&A] [answered]
What is the difference between having a closed form formula and having a system of equations?

**Comment by thelfer** [comment, marked as answer]
With a closed form formula, you don't need to use an integration algorithm, such as Runge-Kutta or theta-scheme. It is conceptually much simplier.

**Reply by superflyer11** [reply]
Does the DefaultDSL @Integrator block do the same thing as the Implicit @Integrator block?

**Reply by thelfer** [reply]
No, not at all:

- In the `Default` DSL, you must compute directly the evolution of the state variable and the value of stress at the end of the time step.
- In the `Implicit` DSL, you have to build a residual and jacobian as a function of the increment of the state variable. A Newton-like algorithm is used to find the zero of this residual.

**Reply by superflyer11** [reply]
Is there any way to automatically calculate the jacobian based on the flow equations or is it necessary to assign them manually?

**Reply by thelfer** [reply]
You can compute the jacobian by finite difference by selecting the algorithm `NewtonRaphson_NumericalJacobian`. See https://github.com/thelfer/tfel/blob/master/mfront/tests/behaviours/T91ViscoplasticBehaviour.mfront

You probably need to specify the perturbation value using `@PerturbationValueForNumericalJacobianComputation`. I would recommand a value between 1e-6 and 1e-8.

**Reply by superflyer11** [reply]
What's the difference between using this and building the jacobian manually? Is there any limitation to this?

**Reply by thelfer** [reply]
Let n be the total number of components of the state variables, computing the numerical jacobian requires 2 * n evaluations.

For instance, in 3D, with a behaviour defining the elastic strain (symmetric tensor) and the equivalent plastic strain (scalar) as state variables, you'll have 7 components. The residual is thus evaluated 15 times per iteration: 1 with the current estimate and 14 to compute the numerical jacobian.

**Reply by superflyer11** [reply]
Hi Thomas,

I am looking at the isotropic von Mises in the gallery here: https://thelfer.github.io/tfel/web/IsotropicLinearHardeningPlasticity.html
[VMHardening_DefaultDSL.mfront.txt](https://github.com/user-attachments/files/17446767/VMHardening_DefaultDSL.mfront.txt)
as well as the equivalent implicit DSL implementation I dug up from a discussion on here previously
[VMHardening_ImplicitDSL_Simo.mfront.txt](https://github.com/user-attachments/files/17446769/VMHardening_ImplicitDSL_Simo.mfront.txt)

I want to extract the plastic strain tensor in mtest,
but I am thinking since the plastic strain was not declared, in the DefaultDSL implementation I added these dirty code to do so. I am not even sure if this is correct:
```
@DSL DefaultDSL;
...

@StateVariable StrainStensor epl;
epl.setGlossaryName("PlasticStrain"); // purely for visualizing data when running in mtest
...
@ProvidesSymmetricTangentOperator;
@Integrator {
   ...
    // using this line to extract the plastic strain
    depl += dp * n;
    ...
}
```
I can indeed extract `epl` but is there a more appropriate way to do extract the plastic strain?


In the implicit implementation, I wanted to do the same thing but I do not know how to do so:
```
@DSL Implicit;
...

// @AuxiliaryStateVariable StrainStensor epl;
// epl.setGlossaryName("PlasticStrain");
...

// integration of constitutive equations
@Integrator {
  ...
  feel += dp * df_dsig;
  fp = f / young;

  // is there a way here or in another block to extract the plastic strain?
  // const auto n = df_dsig;
  // epl = p * n;
}
```

Again is there a way to do the same thing as above?

**Reply by thelfer** [reply]
Hi, @superflyer11
1. For the default DSL, you need to do this:
    ```c++
    epl += dp * n;
    ```
2. For the implicit implementation, declare the plastic strain as an auxiliary state variable and do in `UpdateAuxiliaryStateVariables`:
     ```c++
    epl += deto-deel;
    ```

**Reply by superflyer11** [reply]
Hi @thelfer,

Thanks for the knowledge so far. Both implementations are successful and I've moved on to the Drucker-Prager in it's most basic form.
[DruckerPragerSimple.mfront.txt](https://github.com/user-attachments/files/17450476/DruckerPragerSimple.mfront.txt)
 I wrote this based on the von Mises implicit implementation above and Drucker-Prager implementation on the gallery. Copied to the bottom for convenience.

The test I used for this model, also copied to the bottom for convenience.
[test.py.txt](https://github.com/user-attachments/files/17450503/test.py.txt)

I am making sure the load path (in blue) does not pass through the cone tip

![{DCC534D1-25C4-42AF-8D03-7A3B6437C4E7}](https://github.com/user-attachments/assets/8b83ff76-31d5-45ec-b718-ff4fc759bd54)

The test tells me the plastic surface is not working. Is there anything wrong with the derivatives?

model:
```
@DSL Implicit;
@Behaviour DruckerPragerSimple;
@Date 15 / 10 / 2024;
@Algorithm NewtonRaphson_NumericalJacobian;
@PerturbationValueForNumericalJacobianComputation 1.e-8;
@ModellingHypotheses{".+"};
@Description{
  Simple Drucker Prager with normal plastic flow
  Yield surface defined by:
  "\["
  "  f(p, q) = M_JP * p + \frac{q}{\sqrt{3}} - \frac{M_JP * c'}{\tan(\phi)}"
  "\]"
  Compared to gallery implementation,
  a more elaborate form based on material properties.
  the von Mises norm q is corrected back to sqrtJ2 for convenience
  Definition of M_JP in triaxial compression:
  "\["
  "  M_{JP} = \frac{2 \sqrt{3} \sin(\phi')}{3 - \sin(\phi')}"
  "\]"
  Definition of M_JP in triaxial extension:
  "\["
  "  M_{JP} = \frac{2 \sqrt{3} \sin(\phi')}{3 + \sin(\phi')}"
  "\]"
}


@Epsilon 1.e-10;
@Theta 1.;

@Brick StandardElasticity;
// phi is the internal friction angle
@MaterialProperty real phi;
phi.setEntryName("phi");
// c is cohesion
@MaterialProperty real c;
c.setEntryName("c");

@LocalVariable real lambda;
@LocalVariable real mu;
@LocalVariable bool bpl;
@LocalVariable real MJP;

@StateVariable strain p;
p.setGlossaryName("EquivalentPlasticStrain");
@AuxiliaryStateVariable stress sigeeqtrial;
sigeeqtrial.setEntryName("TrialEquivalentStress"); // purely for visualizing data when running in mtest
@AuxiliaryStateVariable StrainStensor epl;
epl.setGlossaryName("PlasticStrain");


@InitLocalVariables {
  lambda = computeLambda(young, nu);
  mu = computeMu(young, nu);
  MJP = 2 * sqrt(3) * sin(phi) / (3 - sin(phi));
}

// computed at t + theta * dt
@Predictor{
  const auto sigel = computeElasticPrediction();
  const auto pel = trace(sigel) / 3;
  const auto qel = sigmaeq(sigel);
  sigeeqtrial = qel;
  bpl = (MJP * pel + qel / sqrt(3) - MJP * c / tan(phi) > 0);
}

// each iteration at t + theta * dp
@Integrator {
  const auto id = Stensor::Id();
  const auto seps =  real(1.e-12) * young;
  if (bpl) {
    const auto pr = trace(sig) / 3;
    const auto q = sigmaeq(sig);
    // nq is the derivative of the equivalent von Mises stress
    const auto nq = eval(3 * deviator(sig) * (2 * max(q,seps)));
    // df_dsig here is based on the yield function
    // const auto df_dsig   = eval(nq+(tg/3)*id); // this is the form from the mfront gallery
    const auto df_dsig   = eval(MJP * (1/3) * id + nq * (1/(sqrt(3))));
    const auto n = df_dsig; // for learning purposes
    feel = deel - deto + dp * n; // clearer for learning purpose
    fp =  (MJP * pr + q / sqrt(3) - MJP * c / tan(phi))/young;
  }
} // end of @Integrator

@UpdateAuxiliaryStateVariables
{
  epl += deto-deel;
}

// not needed for a single surface (?)
// @AdditionalConvergenceChecks {
//   if (converged){
//     if (bpl) {
//       if (dp < 0) {
//         // desactivating this system
//         converged = bpl = false;
//       }
//     } else {
//       const auto pr = trace(sig) / 3;
//       const auto q = sigmaeq(sig);
//       if (q + pr * tg > d) {
//         converged = false;
//         bpl = true;
//       }
//     }
//   }
// }
```
test:
```
import os
import sys
import numpy as np
import mtest

m = mtest.MTest()
# mtest.setVerboseMode(mtest.VerboseLevel.VERBOSE_QUIET)
m.setMaximumNumberOfSubSteps(20)
m.setModellingHypothesis("Tridimensional")

model = "DruckerPragerSimple"


# Set  material model implementation and path
lib_path = "src/libBehaviour.so"

b = mtest.Behaviour('generic', lib_path, model,'Tridimensional')
m.setBehaviour("generic", lib_path, model)
# in mfront gallery:
# tan(beta) * p + q - R_DP_0 = 0
#R_DP_0 is named d in gallery example

# in my own implementation
# - M_JP * p' + J - M_JP * c' / tan(phi) = 0
# J = q / sqrt(3)
# q = sqrt(3J2)

E = 150
nu = 0.3
# phi is the internal friction angle
phi = np.radians(35)
c = 30

# fitting at triaxial compression: lode angle = -30
# M_JP = 2 * np.sqrt(3) * np.sin(phi) / (3 - np.sin(phi))
# d = M_JP * c / np.tan(phi)

# Loading programme
tMax = 1.0  # s , total time
nTime = 200
ltime = np.linspace(0.0, tMax, nTime)
# Environment parameters
m.setExternalStateVariable("Temperature", 293.15)
# Material parameters
m.setMaterialProperty("YoungModulus", E)
m.setMaterialProperty("PoissonRatio", nu)
m.setMaterialProperty("phi", phi)
m.setMaterialProperty("c", c)

controls = ["stress", "strain"]
control = controls[0]
if control == "stress":
    p_con_0 =  0 # confining pressure
    p_con_1 =  00 # confining pressure
    p_axi_0 = 0
    p_axi_1 = 45
    m.setImposedStress("SXX", {0: p_con_0, 0.02: p_con_0, 1.0: p_con_1})
    m.setImposedStress("SYY", {0: p_con_0, 0.02: p_con_0, 1.0: p_con_1})
    m.setImposedStress("SZZ", {0: p_axi_0, 0.02: p_axi_0, 1.0: p_axi_1})
if control == "strain":
    # Set imposed strains for axial and confining directions
    m.setImposedStrain("EXX", {0: 0, 1.0: 0.00001})
    m.setImposedStrain("EYY", {0: 0, 1.0: 0.00001})
    m.setImposedStrain("EZZ", {0: 0, 1.0: 2.7})


s = mtest.MTestCurrentState()
wk = mtest.MTestWorkSpace()
m.completeInitialisation()
m.initializeCurrentState(s)
m.initializeWorkSpace(wk)


# run sim
for i in range(nTime - 1):
    # print(f"===========Loop {i}===========")
    m.execute(s, wk, ltime[i], ltime[i + 1])

```

**Reply by superflyer11** [reply]
form of the yield surface in my textbook:
![{E9EB8FF4-65C1-41FF-B923-C6BEB634DD80}](https://github.com/user-attachments/assets/e507e1a8-d47f-4359-a98a-21ca0b34e27e)
![{116F627F-C780-47C7-90D1-5D394BBD8E29}](https://github.com/user-attachments/assets/4088b990-7bdd-4b84-a52c-47140f7fbe16)
![{5D9861B5-F707-4C06-9B25-AF9642775530}](https://github.com/user-attachments/assets/ebb67b16-6745-4d72-a8fc-7e50bd272950)
![{E551F3C8-6F5F-4198-B85F-26ACAD262859}](https://github.com/user-attachments/assets/ef9adc8e-54f4-456d-9609-fdfde5b8ad1f)

c and phi are material properties

**Reply by thelfer** [reply]
- It seems that the pressure must have a negative sign.
- Your `python` file misses an `import std`
- the expression of the normal is wrong:
  ```c++
    const auto q = sigmaeq(sig);
    const auto iq = 1 / max(q, seps);
    // nq is the derivative of the equivalent von Mises stress
    const auto nq = eval(3 * deviator(sig) *  (iq / 2));
  ```
  
That being said, if I increase the loading, I reach the plastic surface, but the file does not converge

**Reply by superflyer11** [reply]
Now I have corrected the sign convention for the hydrostatic part and the expression of the normal 
```
@DSL Implicit;
@Behaviour DruckerPragerSimple;
@Date 15 / 10 / 2024;
@Algorithm NewtonRaphson_NumericalJacobian;
@PerturbationValueForNumericalJacobianComputation 1.e-8;
@ModellingHypotheses{".+"};
@Description{
  Simple Drucker Prager with normal plastic flow
  Yield surface defined by:
  "\["
  "  f(p, q) = - M_JP * p + \frac{q}{\sqrt{3}} - \frac{M_JP * c'}{\tan(\phi)}"
  "\]"
  Compared to gallery implementation,
  a more elaborate form based on material properties.
  the von Mises norm q is corrected back to a textbook sqrtJ2 for convenience
  Definition of M_JP in triaxial compression:
  "\["
  "  M_{JP} = \frac{2 \sqrt{3} \sin(\phi')}{3 - \sin(\phi')}"
  "\]"
  Definition of M_JP in triaxial extension:
  "\["
  "  M_{JP} = \frac{2 \sqrt{3} \sin(\phi')}{3 + \sin(\phi')}"
  "\]"
}


@Epsilon 1.e-10;
@Theta 1.;

@Brick StandardElasticity;
// phi is the internal friction angle
@MaterialProperty real phi;
phi.setEntryName("phi");
// c is cohesion
@MaterialProperty real c;
c.setEntryName("c");

@LocalVariable real lambda;
@LocalVariable real mu;
@LocalVariable bool bpl;
@LocalVariable real MJP;

@StateVariable strain p;
p.setGlossaryName("EquivalentPlasticStrain");
@AuxiliaryStateVariable stress sigeeqtrial;
sigeeqtrial.setEntryName("TrialEquivalentStress"); // purely for visualizing data when running in mtest
@AuxiliaryStateVariable StrainStensor epl;
epl.setGlossaryName("PlasticStrain");
@AuxiliaryStateVariable real test;
test.setEntryName("test");



@InitLocalVariables {
  lambda = computeLambda(young, nu);
  mu = computeMu(young, nu);
  MJP = 2 * sqrt(3) * sin(phi) / (3 - sin(phi));
}

// computed at t + theta * dt
@Predictor{
  const auto sigel = computeElasticPrediction();
  const auto pel = trace(sigel) / 3;
  const auto qel = sigmaeq(sigel);
  sigeeqtrial = qel;
  bpl = (- MJP * pel + qel / sqrt(3) - MJP * c / tan(phi) > 0);
  test = - MJP * pel + qel / sqrt(3) - MJP * c / tan(phi);
}

// each iteration at t + theta * dp
@Integrator {
  const auto id = Stensor::Id();
  const auto seps =  real(1.e-12) * young;
  if (bpl) {
    const auto pr = trace(sig) / 3;
    const auto q = sigmaeq(sig);
    // nq is the derivative of the equivalent von Mises stress
    const auto iq = 1 / max(q, seps);
    const auto nq = eval(3 * deviator(sig) *  (iq / 2));
    // const auto nq = eval(3 * deviator(sig) / (2 * max(q,seps)));

    // df_dsig here is based on the yield function
    // const auto df_dsig   = eval(nq+(tg/3)*id); // this is the form from the mfront gallery
    const auto df_dsig   = eval(- MJP * (1/3) * id + (nq /sqrt(3)));
    feel += dp * df_dsig; // clearer for learning purpose
    fp =  (- MJP * pr + q / sqrt(3) - MJP * c / tan(phi))/young;
  }
} // end of @Integrator

@UpdateAuxiliaryStateVariables
{
  epl += deto-deel;
}

// not needed for a single surface (?)
// @AdditionalConvergenceChecks {
//   if (converged){
//     if (bpl) {
//       if (dp < 0) {
//         // desactivating this system
//         converged = bpl = false;
//       }
//     } else {
//       const auto pr = trace(sig) / 3;
//       const auto q = sigmaeq(sig);
//       if (q + pr * tg > d) {
//         converged = false;
//         bpl = true;
//       }
//     }
//   }
// }
```


and the in the log below, what are the numbers in and outside the brackets? I guess the first number is the residual of the first state variable which is eel?
```
...
resolution from 0.879397 to 0.884422
iteration 1 : 0.00444399 0.666598 (-0.229336 -0.229336 0.764455 0 0 0)
iteration 2 : 2.64259e-14 3.96483e-12 (-0.229336 -0.229336 0.764455 0 0 0)
convergence, after 2 iterations, order undefined

resolution from 0.884422 to 0.889447
iteration 1 : 0.00444399 0.666598 (-0.23067 -0.23067 0.768899 0 0 0)
iteration 2 : 2.67084e+12 0.026681 (1.33542e+12 1.33542e+12 -2.67084e+12 0 0 0)

MTest::computeStiffnessMatrixAndResidual : behaviour intregration failed
Dividing time step by two
...
```

Finally I realised this is a perfectly plastic model so I controlled with strain instead, which shows it working

![image](https://github.com/user-attachments/assets/d0831a62-6a84-41b7-8b7f-83eae0b78367)
![image](https://github.com/user-attachments/assets/cdf532d6-29bf-4c00-a68c-ceb20bdccfb7)

**Reply by thelfer** [reply]
> and the in the log below, what are the numbers in and outside the brackets? I guess the first number is the residual of the first state variable which is eel?

`MTest` uses two criterion to check for convergence:

- one on the value of the residual (difference between the imposed stress and the computed stress). Controlled by `@StressEpsilon`.
- one on the stationarity of the strain (difference between two iterations). Controlled by `@StrainEpsilon`.

The values in parenthesis are the components of the total strain.

> Finally I realised this is a perfectly plastic model so I controlled with strain instead, which shows it working

Indeed :)
